File: python/days/day07.py
"""Day 7 of AOC 2022: No Space Left On Device"""

# -------------- Public Functions ------------- #

def day07_pt1(filename: str) -> int:
    """Day 7, part 1: Sum of small directories."""
    with open(filename, 'r', encoding='utf-8') as input_file:
        lines = input_file.read().split('\n')[:-1]

    file_system = FileSystem()
    for line in lines:
        file_system.parse_line(line)

    return sum(file_system.get_small_dirs())

def day07_pt2(filename: str) -> int:
    """Day 7, part 2: Smallest directory to delete."""
    with open(filename, 'r', encoding='utf-8') as input_file:
        lines = input_file.read().split('\n')[:-1]

    file_system = FileSystem()
    for line in lines:
        file_system.parse_line(line)

    return file_system.get_dir_size(file_system.min_dir_to_remove())

# ------------------ Classes ------------------ #

class FileSystem:
    """Class for keeping track of directory traversal."""
    _curr_dir: str
    _directories: dict[str]
    _total_space: int

    def __init__(self, total_space=70_000_000):
        self._curr_dir = '/'
        self._directories = {'/': 0}
        self._total_space = total_space

    def parse_line(self, line: str):
        """Parses a line from the inputfile."""
        match line.split():
            case ['$', 'cd', dir_name]:
                self._change_dir(dir_name)
            case ['$', 'ls']:
                pass
            case ['dir', dir_name]:
                self._add_dir(dir_name)
            case [size, _file_name]:
                self._add_file(int(size))
            case _:
                raise Exception(f'This should not occur...\n\t{line.split()}')

    def get_small_dirs(self, size_limit=100_000) -> list[int]:
        """Get a list of all the small directories."""
        small_dirs = []

        for directory in self._directories.values():
            if directory <= size_limit:
                small_dirs.append(directory)

        return small_dirs

    def get_dir_size(self, dir_name: str) -> int:
        """Get the size of a directory, given its name."""
        return self._directories[dir_name] # Bad error handling

    def min_dir_to_remove(self, space_needed=30_000_000) -> str:
        """Find the smallest directory to remove to get the space needed."""
        space_to_remove = space_needed - self.space_left()

        return min([
                (name, size) for (name, size) in self._directories.items()
                if size >= space_to_remove
            ],
            key=lambda tpl: tpl[1]
        )[0]

    def space_left(self) -> int:
        """Get how much space is left in the file system."""
        return self._total_space - self._directories['/']

    def _change_dir(self, dir_name: str):
        match dir_name:
            case '..':
                if self._curr_dir != '/':
                    self._curr_dir = '/'.join(self._curr_dir.split('/')[:-2]) + '/'
            case '/':
                self._curr_dir = '/'
            case new_dir:
                self._curr_dir += new_dir + '/'

    def _add_dir(self, dir_name: str):
        new_dir = self._curr_dir + dir_name + '/'
        if not new_dir in self._directories:
            self._directories[new_dir] = 0

    def _add_file(self, file_size: int):
        curr = '/'
        for directory in self._curr_dir.split('/')[1:]:
            self._directories[curr] += file_size
            curr += directory + '/'


# ----------------- Constants ----------------- #


# ------------- Private Functions ------------- #


# ------------------- Script ------------------ #

if __name__ == '__main__':
    INPUT_FILENAME = '../inputs/07.txt'
    print(f'Day 7, pt1: {day07_pt1(INPUT_FILENAME)}')
    print(f'Day 7, pt2: {day07_pt2(INPUT_FILENAME)}')

File: typescript/src/days/day07.ts
// Import NodeJS helpers
import fs from 'fs';

// Import types
import InputFile from '../types/InputFile';

// ------------------ Types ------------------ //

type FileSystem = {
  currDir: string,
  directories: Directories,
};

// type Directories = {
//   [k: string]: number,
// };

type Directories = Map<string, number>;

// ----------------- Helpers ----------------- //

const newFileSystem = (): FileSystem => {
  const currDir = '/';
  const directories = new Map([['/', 0]]);

  return {
    currDir,
    directories,
  };
};

const changeDir = (fileSystem: FileSystem, dirName: string): FileSystem => {
  const {
    currDir,
    directories,
  } = fileSystem;
  if (dirName === '..' && currDir !== '/') {
    const nextDir = currDir
      .split('/')
      .slice(0, -2)
      .join('/');
    return {
      currDir: `${nextDir}/`,
      directories,
    };
  }
  if (dirName === '/') {
    return {
      currDir: '/',
      directories,
    };
  }
  return {
    currDir: `${currDir}${dirName}/`,
    directories,
  };
};

const addDir = (fileSystem: FileSystem, dirName: string): FileSystem => {
  const {
    currDir,
    directories,
  } = fileSystem;

  directories.set(`${currDir}${dirName}/`, 0);

  return {
    currDir,
    directories,
  };
};

const addFile = (fileSystem: FileSystem, fileSize: number): FileSystem => {
  const {
    currDir,
    directories,
  } = fileSystem;

  const dirSteps = currDir
    .replaceAll('/', '/,')
    .split(',')
    .slice(0, -1);

  dirSteps.reduce(
    (dir, next) => {
      const nextDir = dir + next;
      const currSize = directories.get(nextDir);
      directories.set(nextDir, currSize + fileSize);
      return nextDir;
    },
    '',
  );

  return {
    currDir,
    directories,
  };
};

const parseLine = (fileSystem: FileSystem, line: string): FileSystem => {
  const cmds = line.split(' ');

  if (cmds[0] === '$' && cmds[1] === 'cd') {
    return changeDir(fileSystem, cmds[2]);
  }
  if (cmds[0] === 'dir') {
    return addDir(fileSystem, cmds[1]);
  }
  const fileSize = Number.parseInt(cmds[0], 10);
  if (!Number.isNaN(fileSize)) {
    return addFile(fileSystem, fileSize);
  }

  return fileSystem;
};

const sumOfSmall = (fileSystem: FileSystem, sizeLimit = 100_000): number => {
  const { directories } = fileSystem;

  let sum = 0;
  for (const dirSize of directories.values()) {
    if (dirSize <= sizeLimit) sum += dirSize;
  }
  return sum;
};

const minDirToDelete = (fileSystem: FileSystem, totalSpace = 70_000_000, spaceNeeded = 30_000_000): number => {
  const { directories } = fileSystem;

  const spaceTaken = directories.get('/');

  const toRemove = spaceNeeded - (totalSpace - spaceTaken);

  let min = totalSpace;
  for (const dirSize of directories.values()) {
    if (dirSize >= toRemove && dirSize < min) min = dirSize;
  }

  return min;
};

// ------------- Main Functions -------------- //

/**
 * Day 7 Pt1 of Advent of Code 2022: No Space Left On Device!
 * Get the sum of the small directories.
 * @author Benedikt Arnarsson
 */
const pt1 = async (filename: InputFile) => (
  fs.promises
    .readFile(filename)
    .then((buf) => buf.toString())
    .then((str) => str.split('\n'))
    .then((lines) => lines.reduce(parseLine, newFileSystem()))
    .then(sumOfSmall)
);

/**
 * Day 7 Pt2 of Advent of Code 2022: No Space Left On Device!
 * Get the size of the smallest directory to delete.
 * @author Benedikt Arnarsson
 */
const pt2 = async (filename: InputFile) => (
  fs.promises
    .readFile(filename)
    .then((buf) => buf.toString())
    .then((str) => str.split('\n'))
    .then((lines) => lines.reduce(parseLine, newFileSystem()))
    .then(minDirToDelete)
);

const day07 = [pt1, pt2];

export default day07;
